### PR TITLE
Honda Bosch Radarless: experimental longitudinal support

### LINKED ISSUE #25355: Honda Bosch Radarless: add longitudinal support behind toggle
The 22 Civic is our first supported car on the radarless Honda Bosch platform. openpilot doens't control longitudinal yet, due to our current longitudinal policy not working well without radar. However, longitudinal control should be supported behind the [DisableRadar toggle](https://github.com/commaai/openpilot/blob/4bfbd7ced822e3bb52fe7f6f446fad65bcb215f4/selfdrive/ui/qt/offroad/settings.cc#L81), so it's ready once we do have a good radarless policy.

### PR DESCRIPTION
**Pull request by royjr**
Possible pull for #25355. Works great except stops way too close to the lead car.
[See panda PR](https://github.com/commaai/panda/pull/1008)

### CODE REVIEW

**Comment by sshane** [comment]
Can you provide some routes with rlogs uploaded in stop and go traffic and highway speeds? I'd like to confirm we understand the the accel interface and it follows our request well enough

File: selfdrive/car/honda/interface.py [hunk 1]
```diff
@@ -36,12 +36,7 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), car_fw=[], disabl
     if candidate in HONDA_BOSCH:
       ret.safetyConfigs = [get_safety_config(car.CarParams.SafetyModel.hondaBosch)]
       ret.radarOffCan = True
-
-      if candidate not in HONDA_BOSCH_RADARLESS:
-        # Disable the radar and let openpilot control longitudinal
-        # WARNING: THIS DISABLES AEB!
-        ret.openpilotLongitudinalControl = disable_radar
-
+      ret.openpilotLongitudinalControl = disable_radar
```
Comment (sshane) [inline selfdrive/car/honda/interface.py:39]:
```suggestion
      # Disable the radar and let openpilot control longitudinal
      # WARNING: THIS DISABLES AEB!
      # If Bosch radarless, this blocks ACC messages
      ret.openpilotLongitudinalControl = disable_radar
```

**Comment by sshane** [comment]
Also should figure out what sends AEB and FCW so we can make sure those are never blocked (I think the camera does it?)

**Comment by royjr** [comment]
I assume FCW is the warning before AEB takes over and starts braking. I believe the camera sends those messges from our last [findings](https://github.com/commaai/opendbc/pull/651). I was never able to trigger AEB on my car confirm the AEB messages (AEB_BRAKING, AEB_STATUS)

File: selfdrive/car/honda/hondacan.py [hunk 2]
```diff
@@ -51,33 +51,40 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
...
-    "SET_TO_FF": 0xff,
-    "SET_TO_75": 0x75,
-    "SET_TO_30": 0x30,
-  }
-  commands.append(packer.make_can_msg("ACC_CONTROL_ON", bus, acc_control_on_values))
-
+  if car_fingerprint in HONDA_BOSCH_RADARLESS:
+    acc_control_values = {
+      "CONTROL_ON": control_on,
+      "CONTROL_OFF": control_off,
```
Comment (sshane) [inline selfdrive/car/honda/hondacan.py:64]: Did you figure out which bit controls actuation, if any? Or do both?

File: selfdrive/car/honda/hondacan.py [hunk 3]
```diff
@@ -51,33 +51,40 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
...
+    commands.append(packer.make_can_msg("ACC_CONTROL", bus, acc_control_values))
+  else:
+    acc_control_values = {
+      # setting CONTROL_ON causes car to set POWERTRAIN_DATA->ACC_STATUS = 1
+      "CONTROL_ON": control_on,
+      "GAS_COMMAND": gas_command,  # used for gas
+      "ACCEL_COMMAND": accel_command,  # used for brakes
+      "BRAKE_LIGHTS": braking,
+      "BRAKE_REQUEST": braking,
+      "STANDSTILL": standstill,
```
Comment (sshane) [inline selfdrive/car/honda/hondacan.py:76]: It might be good to find any standstill bits in the message so we can set those if needed.

**Comment by sshane** [comment]
- [ ] Figure out what the stock system requests to hold a standstill, and does it set any bits?

File: selfdrive/car/honda/hondacan.py (hunk 2 above)
Comment (royjr) [reply selfdrive/car/honda/hondacan.py:64]: It looks like you only need CONTROL_ON just like the comment says. Tried without setting either CONTROL and with just CONTROL_OFF, but none moved the car or changed ACC_STATUS.

1a5d045d2c531a6d|2022-08-08--18-41-25
1a5d045d2c531a6d|2022-08-08--18-41-25

**Comment by royjr** [comment]
1a5d045d2c531a6d|2022-09-01--13-46-28
1a5d045d2c531a6d|2022-09-01--13-36-53

**Comment by royjr** [comment]
tux: ```undefined signal STANDSTILL - 456```
There is no STANDSTILL in ACC_CONTROL for this car.

**Comment by sshane** [comment]
Should be fixed now

**Comment by sshane** [comment]
Should be fixed now. Can you tell me a few things?

- With stock ACC, do you need to press the resume button or press the accelerator to resume from a stop?
- Before the STANDSTILL addition on openpilot ACC, do you need to press the resume button or press the accelerator to resume from a stop?

**Comment by royjr** [comment]
**Stock ACC** - _needs resume button_
**OP + Stock ACC** - _does not need resume button, but sometimes doesnt resume automatically (once in a blue)_
**OP + OP ACC** - _does not need resume button, but short terms tests may not be conclusive based on above_
**OP + OP ACC + STANDSTILL addition** - _does not need resume button, but short terms tests may not be conclusive based on above_

**Comment by royjr** [comment]
@dmdreon It definitely works, just has a few ui kinks left to iron out.
- Mismatching speed when setting speed below 25
- Dashboard blinks speed sometimes
- Dashboard "stopped" remains after resume

File: selfdrive/car/honda/hondacan.py [hunk 4]
```diff
@@ -111,13 +119,17 @@ def create_ui_commands(packer, CP, enabled, pcm_speed, hud, is_metric, acc_hud,
...
       'IMPERIAL_UNIT': int(not is_metric),
       'HUD_LEAD': 2 if enabled and hud.lead_visible else 1 if enabled else 0,
       'SET_ME_X01_2': 1,
     }
 
+    if CP.carFingerprint in HONDA_BOSCH_RADARLESS:
+      acc_hud_values['ENABLE_MINI_CAR'] = 1 if enabled else 0
+    else:
+      acc_hud_values['ENABLE_MINI_CAR'] = 1
+
```
Comment (sshane) [inline selfdrive/car/honda/hondacan.py:132]: this looks like a general improvement/clean up for all Honda. can we make it another PR? is it critical to this PR?

File: selfdrive/car/honda/hondacan.py [hunk 5]
```diff
@@ -58,27 +58,35 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
...
-  }
-  commands.append(packer.make_can_msg("ACC_CONTROL_ON", bus, acc_control_on_values))
-
+  if car_fingerprint in HONDA_BOSCH_RADARLESS:
+    control_on = 1 if enabled else 0
+    acc_control_values = {
+      "CONTROL_ON": control_on,
+      "CONTROL_OFF": 1,  # TODO: not sure this signal is needed
+      "ACCEL_COMMAND": accel_command,
+      "STANDSTILL": standstill,
```
Comment (sshane) [inline selfdrive/car/honda/hondacan.py:67]: the signals for this message name defined already below, are they common at all in the new acc control msg? are there weird constant setmes?

File: selfdrive/car/honda/carcontroller.py [hunk 6]
```diff
@@ -127,7 +128,27 @@ def __init__(self, dbc_name, CP, VM):
...
+      if CC.longActive and stopping:
+        if self.last_stopping_frame == 0:
+          self.last_stopping_frame = self.frame
+        if self.frame - self.last_stopping_frame > 400:
+          hud_v_cruise = 252
+        else:
+          if CS.is_metric:
+            hud_v_cruise = hud_control.setSpeed * CV.MS_TO_KPH if hud_control.speedVisible else 255
+          else:
+            hud_v_cruise = hud_control.setSpeed * CV.MS_TO_MPH if hud_control.speedVisible else 255
```
[the remark this answers is not in the document]
Comment (royjr) [reply selfdrive/car/honda/carcontroller.py]: Slightly better?
```
if self.CP.carFingerprint in HONDA_BOSCH_RADARLESS:
  if CC.longActive and (actuators.longControlState == LongCtrlState.stopping):
    if self.last_stopping_frame == 0:
      self.last_stopping_frame = self.frame
    if self.frame - self.last_stopping_frame > 400:
      hud_v_cruise = 252
    else:
      hud_v_cruise = hud_control.setSpeed * (CV.MS_TO_KPH if CS.is_metric else CV.MS_TO_MPH) if hud_control.speedVisible else 255
  else:
    self.last_stopping_frame = 0
    hud_v_cruise = hud_control.setSpeed * (CV.MS_TO_KPH if CS.is_metric else CV.MS_TO_MPH) if hud_control.speedVisible else 255
else:
  hud_v_cruise = hud_control.setSpeed * CV.MS_TO_MPH if hud_control.speedVisible else 255
```

[the remark this answers is not in the document]
Comment (royjr) [reply selfdrive/car/honda/carcontroller.py]: Maybe even better?
```
if self.CP.carFingerprint in HONDA_BOSCH_RADARLESS:
  hud_v_cruise = hud_control.setSpeed * (CV.MS_TO_KPH if CS.is_metric else CV.MS_TO_MPH) if hud_control.speedVisible else 255
  if CC.longActive and (actuators.longControlState == LongCtrlState.stopping):
    if self.last_stopping_frame == 0:
      self.last_stopping_frame = self.frame
    if self.frame - self.last_stopping_frame > 400:
      hud_v_cruise = 252
  else:
    self.last_stopping_frame = 0
else:
  hud_v_cruise = hud_control.setSpeed * CV.MS_TO_MPH if hud_control.speedVisible else 255
```

File: selfdrive/car/honda/hondacan.py (hunk 5 above)
Comment (royjr) [reply selfdrive/car/honda/hondacan.py]: Some are common, but might have different values like CONTROL_ON. I moved common messages out of the if statement.

File: selfdrive/car/honda/carcontroller.py [hunk 7]
```diff
@@ -127,7 +128,18 @@ def __init__(self, dbc_name, CP, VM):
...
-    hud_v_cruise = hud_control.setSpeed * CV.MS_TO_KPH if hud_control.speedVisible else 255
+
+    hud_v_cruise = hud_control.setSpeed * CV.MS_TO_MPH if hud_control.speedVisible else 255
+    if self.CP.carFingerprint in HONDA_BOSCH_RADARLESS:
+      conversion = CV.MS_TO_KPH if CS.is_metric else CV.MS_TO_MPH
+      hud_v_cruise = hud_control.setSpeed * conversion if hud_control.speedVisible else 255
+      if CC.longActive and actuators.longControlState == LongCtrlState.stopping:
+        self.stopping_counter += 1
+        if self.stopping_counter > 400:
+          hud_v_cruise = 252
```
Comment (sshane) [inline selfdrive/car/honda/carcontroller.py:139]: can you make a global variable with a comment for this magic number of what it does?

Comment (royjr) [reply selfdrive/car/honda/carcontroller.py:142]: I'm assuming the magic number in reference is 400 and not 252, or should we make both variables?

Comment (sshane) [reply selfdrive/car/honda/carcontroller.py:142]: 400 is very clearly 4 seconds, I meant what that set speed did

Comment (sshane) [reply selfdrive/car/honda/carcontroller.py:142]: Does showing stopped change any behavior like turn off the engine?

Comment (royjr) [reply selfdrive/car/honda/carcontroller.py]: Silly me, 400 wasn't 4 seconds to me because it would start counting down before it came to a stop. 252 is the stopped message which we now don't need.

**Comment by royjr** [comment]
A few things come to mind..

- There was no conversion logic before, it was always CV.MS_TO_KPH which doesn't work for the civic. This might break non-radarless cars. Should we only apply the conversion to radarless?
- Do we still need the gas_override in the hondacan ui function or can we just do that check in carcontroller since both could set to the same 255 value?

**Comment by sshane** [comment]
Yes, how you had it before with the radar less check was fine. We can bring out the override check too

**Comment by royjr** [comment]
done, unless you wanted to move the override check to another PR since thats new to other hondas

**Comment by sshane** [comment]
@royjr when this is merged, can you open new PRs for the follow?:

- idlestop timer, if stock uses a timer and doesn't set it immediately or statically
- match stock hide speed behavior on override

File: selfdrive/car/honda/hondacan.py [hunk 8]
```diff
@@ -51,34 +56,42 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
...
-    "CONTROL_ON": enabled,
-    "SET_TO_FF": 0xff,
-    "SET_TO_75": 0x75,
-    "SET_TO_30": 0x30,
-  }
-  commands.append(packer.make_can_msg("ACC_CONTROL_ON", bus, acc_control_on_values))
+  if car_fingerprint in HONDA_BOSCH_RADARLESS:
+    acc_control_values.update({
+      "CONTROL_ON": enabled,
+      "IDLESTOP_ALLOW": 1,  # allows idle stop
```
Comment (sshane) [inline selfdrive/car/honda/hondacan.py:74]:
```suggestion
      "IDLESTOP_ALLOW": 0,  # disallows idle stop
```

File: selfdrive/car/honda/hondacan.py [hunk 9]
```diff
@@ -51,34 +56,42 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
   bus = get_pt_bus(car_fingerprint)
   min_gas_accel = CarControllerParams.BOSCH_GAS_LOOKUP_BP[0]
 
-  control_on = 5 if enabled else 0
   gas_command = gas if active and accel > min_gas_accel else -30000
```
Comment (sshane) [inline selfdrive/car/honda/hondacan.py:59]:
```suggestion
  control_on = 5 if enabled else 0
  gas_command = gas if active and accel > min_gas_accel else -30000
```
